src/components/nano/ModalChangeAddress.js

```diff
@@ -0,0 +1,171 @@
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import React, { useEffect, useRef, useState } from 'react';
+import { t } from 'ttag';
+import $ from 'jquery';
+import PropTypes from 'prop-types';
+import AddressList from '../../components/AddressList';
+import { CHANGE_ADDRESS_LIST_COUNT } from '../../constants';
+import { connect } from 'react-redux';
+import { useDispatch, useSelector } from 'react-redux';
+import { editAddressNC } from '../../actions/index';
+import { getGlobalWallet } from '../../modules/wallet';
+
+/**
+ * Component that shows a modal to change the address of a Nano Contract
+ *
+ * @memberof Components
+ */
+function ModalChangeAddress(props) {
+  const dispatch = useDispatch();
+  const wallet = getGlobalWallet();
+  const nanoContracts = useSelector(state => state.nanoContracts);
+  const oldAddress = nanoContracts[props.nanoContractID].address;
+
+  // {number} step The modal has 3 steps: 0 is the warning screen, 1 is the address list and 2 is the confirmation
+  const [step, setStep] = useState(0);
+  // {Object} newAddress New nano contract selected address data {'address': string, 'index': number}
+  const [newAddress, setNewAddress] = useState(null);
+  // {number} oldAddressIndex the index of the old selected address. Need to be as a state because of the async/await call to the wallet method
+  const [oldAddressIndex, setOldAddressIndex] = useState(null);
+
+  useEffect(() => {
+    $('#changeAddressModal').modal('show');
+    $('#changeAddressModal').on('hidden.bs.modal', (e) => {
+      setStep(0);
+      // We always need to call on close when using global context modal
+      props.onClose();
+    })
+
+    async function getAndSetOldAddressIndex() {
+      const oldAddressIndex = await wallet.getAddressIndex(oldAddress);
+      setOldAddressIndex(oldAddressIndex);
+    }
+
+    getAndSetOldAddressIndex();
+
+    return () => {
+      $('#changeAddressModal').modal('hide');
+      // Removing all event listeners
+      $('#changeAddressModal').off();
+    };
+
+  }, []);
+
+  /**
+   * Called when user clicks to continue after reading the warning message in step 0
+   *
+   * @param {Object} e Event emitted on click
+   */
+  const goToStep1 = (e) => {
+    e.preventDefault();
+    setStep(1);
+  }
+
+  /**
+   * Called when user clicks in one of the addresses in step 1
+   *
+   * @param {string} address Address clicked
+   */
+  const onAddressClick = async (address) => {
+    const newAddressIndex = await wallet.getAddressIndex(address);
+    setNewAddress({ address, index: newAddressIndex});
+    setStep(2);
+  }
+
+  /**
+   * Called when user clicks to change the address in step 2
+   *
+   * @param {Object} e Event emitted on click
+   */
+  const executeChange = async (e) => {
+    e.preventDefault();
+    dispatch(editAddressNC(props.nanoContractID, newAddress.address));
+    await wallet.storage.updateNanoContractRegisteredAddress(props.nanoContractID, newAddress.address);
```

Comment:
question: Shouldn't the `editAddressNC` action also update the wallet storage? It seems one line execution wouldn't make much sense without the other.

It looks to me like the kind of work to be done by a saga.

Reply:
I don't think it makes sense to have storage update in the redux action but moving to a saga makes sense. Done https://github.com/HathorNetwork/hathor-wallet/pull/623/commits/a68294c5e7e2610ea398580b503812eef17a9590